### PR TITLE
External weight change

### PR DESCRIPTION
**Pull request by rowleya**
Allow an external weight change from a different connection.  This is a bit like STDP with neuromodulation, but much simpler.  A projection can be "WeightChangable" which is the "STDP" part.  This can then also have a "WeightChanger" projection which is like the neuromodulation; this takes the WeightChangable projection as a parameter, and the source Population must have the same number of neurons as the Projection it affects, and must of course target the same target Population and receptor type (this is checked).  The WeightChanger connections are then set up so that they affect a specific pre-post pair of neurons.  When the Pre-Population of the WeightChanger spikes, the weight is changed by the specified amount (which can be positive or negative) on any connection that matches the same pre-post pairing in the target.

Tested with:
 - https://github.com/SpiNNakerManchester/IntegrationTests/pull/293

### CODE REVIEW

File: spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+        "__weight_max",
+
+        # The minimum weight
+        "__weight_min",
+
+        # The map of synapse information to index
+        "__synapse_info_to_index",
+
+        # The next index to use for the next projection
+        "__next_index")
```
Comment (Christian-B) [inline spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py:78]: Not needed!

File: spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+        """
+        return self.__weight_min
+
+    def get_synapse_info_index(self, synapse_info: SynapseInformation) -> int:
+        """ Get the row offset for the given synapse information.  Each synapse
+            information has a unique row offset which then allows for multiple
+            connections to be identified and kept separate.
+        """
+        if synapse_info not in self.__synapse_info_to_index:
+            self.__synapse_info_to_index[synapse_info] = self.__next_index
```
Comment (Christian-B) [inline spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py:114]: self.__synapse_info_to_index[synapse_info] = len(self.__synapse_info_to_index) is easier and safer

File: spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+            weight: _In_Types = StaticSynapse.default_parameters['weight'],
+            delay: _In_Types = None):
+        """
+        :param float weight:
+        :param delay: Use ``None`` to get the simulator default minimum delay.
+        :type delay: float or None
+        """
+        super().__init__(delay=delay, weight=weight)
+        self.__weight_max = weight_max
+        self.__weight_min = weight_min
```
Comment (Christian-B) [inline spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py:92]: max and min should be validated!
1. Make sure max is not smaller
2. make sure they are values the code can handle

File: spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py (hunk 2 above)
Comment (rowleya) [reply spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py:114]: But that is incorrect - the index is based on the number of atoms previously added, not the number of connections

File: spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+            synapse_weight_scales * global_weight_scale *
+            self.__weight_min)
+        max_weights = (
+            synapse_weight_scales * global_weight_scale *
+            self.__weight_max)
+        weights = numpy.dstack((min_weights, max_weights)).flatten().astype(
+            uint16).view(uint32)
+        spec.write_array(weights)
+
+    def get_n_words_for_plastic_connections(self, n_connections):
```
Comment (Christian-B) [inline spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py:186]: overrides needed

**Comment by Christian-B** [comment]
We need a test where some changer spikes happen before the changable gets a spike

File: spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+    from spynnaker.pyNN.models.neuron.synapse_dynamics.types import (
+        ConnectionsArray)
+    from spynnaker.pyNN.models.neuron.synapse_io import MaxRowInfo
+    from .abstract_synapse_dynamics import AbstractSynapseDynamics
+
+# How large are the time-stamps stored with each event
+TIME_STAMP_BYTES = BYTES_PER_WORD
+
+# The targets of neuromodulation
+NEUROMODULATION_TARGETS = {
```
Comment (Christian-B) [inline spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py:57]: why is this here?

Comment (rowleya) [reply spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py]: Removed

File: spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+if TYPE_CHECKING:
+    from spynnaker.pyNN.models.neural_projections import (
+        ProjectionApplicationEdge, SynapseInformation)
+    from spynnaker.pyNN.models.neuron.synapse_dynamics.types import (
+        ConnectionsArray)
+    from spynnaker.pyNN.models.neuron.synapse_io import MaxRowInfo
+    from .abstract_synapse_dynamics import AbstractSynapseDynamics
+
+# How large are the time-stamps stored with each event
+TIME_STAMP_BYTES = BYTES_PER_WORD
```
[the remark this answers is not in the document]
Comment (rowleya) [reply spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py]: Removed

File: neural_modelling/src/neuron/plasticity/weight_change/post_events_with_weight_change.h [hunk 7]
```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2017 The University of Manchester
```
Comment (Christian-B) [inline neural_modelling/src/neuron/plasticity/weight_change/post_events_with_weight_change.h:2]: 2024

File: neural_modelling/makefiles/neuron/IF_curr_exp_weight_change/Makefile [hunk 8]
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) 2014 The University of Manchester
```
Comment (Christian-B) [inline neural_modelling/makefiles/neuron/IF_curr_exp_weight_change/Makefile:1]: 2024

File: neural_modelling/src/neuron/plasticity/weight_change/synapse_dynamics_external_weight_change.c [hunk 9]
```diff
@@ -0,0 +1,274 @@
+/*
```
Comment (Christian-B) [inline neural_modelling/src/neuron/plasticity/weight_change/synapse_dynamics_external_weight_change.c:1]: 2024

File: neural_modelling/makefiles/neuron/IF_curr_exp_weight_change/Makefile (hunk 8 above)
Comment (rowleya) [inline neural_modelling/makefiles/neuron/IF_curr_exp_weight_change/Makefile:1]:
```suggestion
# Copyright (c) 2024 The University of Manchester
```

File: neural_modelling/src/neuron/plasticity/weight_change/post_events_with_weight_change.h (hunk 7 above)
Comment (rowleya) [inline neural_modelling/src/neuron/plasticity/weight_change/post_events_with_weight_change.h:2]:
```suggestion
 * Copyright (c) 2024 The University of Manchester
```

File: neural_modelling/src/neuron/plasticity/weight_change/synapse_dynamics_external_weight_change.c [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+ * Copyright (c) 2016 The University of Manchester
```
Comment (rowleya) [inline neural_modelling/src/neuron/plasticity/weight_change/synapse_dynamics_external_weight_change.c:2]:
```suggestion
 * Copyright (c) 2024 The University of Manchester
```

File: neural_modelling/src/synapse_expander/matrix_generators/matrix_generator_weight_changer.h [hunk 11]
```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2017 The University of Manchester
```
Comment (rowleya) [inline neural_modelling/src/synapse_expander/matrix_generators/matrix_generator_weight_changer.h:2]:
```suggestion
 * Copyright (c) 2024 The University of Manchester
```

File: neural_modelling/src/neuron/plasticity/weight_change/post_events_with_weight_change.h [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! \file
+//! \brief Post-synaptic events
+#ifndef _POST_EVENTS_H_
```
Comment (Christian-B) [inline neural_modelling/src/neuron/plasticity/weight_change/post_events_with_weight_change.h:19]: Used in three different h files!

File: spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py [hunk 13]
```diff
@@ -0,0 +1,388 @@
+# Copyright (c) 2015 The University of Manchester
```
Comment (rowleya) [inline spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changable.py:1]:
```suggestion
# Copyright (c) 2024 The University of Manchester
```

File: spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changer.py [hunk 14]
```diff
@@ -0,0 +1,309 @@
+# Copyright (c) 2021 The University of Manchester
```
Comment (rowleya) [inline spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_weight_changer.py:1]:
```suggestion
# Copyright (c) 2024 The University of Manchester
```

File: neural_modelling/src/neuron/plasticity/weight_change/post_events_with_weight_change.h [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+//---------------------------------------
+// Macros
+//---------------------------------------
+//! Maximum number of pre-synaptic events per post neuron
+#define MAX_EVENTS 16
+
+typedef struct update_post_trace_t {
+
+    //! The new weight to set
```
Comment (Christian-B) [inline neural_modelling/src/neuron/plasticity/weight_change/post_events_with_weight_change.h:38]: Comment name mismatch! is is the new weight to set or the amount to change?

File: spynnaker_integration_tests/test_various/test_weight_changer.py [hunk 16]
```diff
@@ -0,0 +1,184 @@
+# Copyright (c) 2017 The University of Manchester
```
Comment (rowleya) [inline spynnaker_integration_tests/test_various/test_weight_changer.py:1]:
```suggestion
# Copyright (c) 2024 The University of Manchester
```

File: neural_modelling/src/neuron/plasticity/weight_change/post_events_with_weight_change.h (hunk 12 above)
Comment (rowleya) [reply neural_modelling/src/neuron/plasticity/weight_change/post_events_with_weight_change.h:19]: Probably not a big problem.  Arguably, they can't be used at the same time... could be fixed if that is somehow needed later!
